### PR TITLE
[Scheduler] Fix spillback is done deterministically.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Before: Choose the node with the min_violation while we iterate nodes in the cluster with the same order always.
Now: Choose the best nodes with the same min violations and choose one of them randomly.

Alternative: Construct a random index array that is equal to len(nodes) and iterate them. I decided to go with the current approach because we will need to be able to choose multiple best nodes anyway for locality aware scheduling. 

Need to see the CI result first.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_multi_node_3.py
Comment (wuisawesome) [inline python/ray/tests/test_multi_node_3.py:411]: Since we really care about uniform spilling behavior, can we make this test a bit more strict? For example `test_actor_advanced.py::test_actor_load_balancing` actually makes sure that the distribution is relatively uniform in the 0 cpus actor case (in fact, maybe we can just copy that test case and make the actors 1 cpu?)

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:229]: Why does the 0 cpu case need to be modified at all? This change looks like it will results in the same behavior.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:277]: Github won't let me add this comment on line 264, but can you add a TODO before the placement hint case that says it isn't sampled at uniform random?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:289]: I think you can remove the TODO now since you've introduced the better load balancing behavior.

**Review by wuisawesome — COMMENTED** [review]
Mostly looks good, btw, do you think it's possible to run the distributed scalability envelope or many_tasks release tests against this to prove that it improves the e2e runtime on a task workload?

**Comment by rkooo567** [comment]
@wuisawesome Is this a hard requirement to merge this PR, or I can do that after merging the PR? 

> Mostly looks good, btw, do you think it's possible to run the distributed scalability envelope or many_tasks release tests against this to prove that it improves the e2e runtime on a task workload?

**Comment by wuisawesome** [comment]
I guess we don't need to block merging the PR on running the release tests, but it would be nice to run it sooner rather than later since there are a lot of potential side effects.

**Comment by rkooo567** [comment]
@wuisawesome Got it. That makes a ton of sense. I can try running test many tasks with smaller number of nodes. Which distirubted scalability envelop tests do you recommend to run?

**Comment by wuisawesome** [comment]
I think the 10k task submission time will be most interesting.

**Comment by rkooo567** [comment]
@wuisawesome You expect this to be faster right?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (rkooo567) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:279]: Done.

File: python/ray/tests/test_multi_node_3.py
Comment (rkooo567) [reply python/ray/tests/test_multi_node_3.py:413]: Good idea. Done. Btw, I created 10 actors in 15 CPUs clusters, and the min count is 2. Does it look reasonable? (3 is pretty hard to achieve).

File: python/ray/tests/test_multi_node_3.py
Comment (wuisawesome) [inline python/ray/tests/test_multi_node_3.py:411]:
```suggestion
    # Make sure tasks are spilled back non-deterministically
```

File: python/ray/tests/test_multi_node_3.py
Comment (wuisawesome) [reply python/ray/tests/test_multi_node_3.py:413]: Hmm a little worried about flakiness with such small numbers and a random test. What about if we just have 2 nodes with 100 CPUs each, and launch 100 actors. Then if we can assert that # spilled to node 1 > 35 with 0.15% flakiness (with 2 nodes, we these parameters we can model spillback with IID coin flips).

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:229]: Not a big deal, just wanted to make sure I understood correctly. (though now it looks like we could just return immediately instead of breaking...).

File: src/ray/core_worker/test/memory_store_test.cc
Comment (wuisawesome) [inline src/ray/core_worker/test/memory_store_test.cc:47]: Is this intentional?

File: python/ray/tests/test_multi_node_3.py
Comment (rkooo567) [reply python/ray/tests/test_multi_node_3.py:413]: Actually, 100 CPUs are not ideal for unit tests environment (machine has only 2 CPUs). What about we merge it and I can adjust the values if it turns out to be flaky?

File: src/ray/core_worker/test/memory_store_test.cc
Comment (rkooo567) [reply src/ray/core_worker/test/memory_store_test.cc:47]: Yes. I found this test was failing on Mac always for some reason.

File: python/ray/tests/test_multi_node_3.py
Comment (wuisawesome) [reply python/ray/tests/test_multi_node_3.py:413]: In the 2 node case, std dev = $$ sqrt{N} / 2 $$. What do you think the largest number of tasks we can do in a unit test is? I think it's pretty likely that this becomes flakey.

Comment (rkooo567) [reply python/ray/tests/test_multi_node_3.py:413]: oh actually, I just realized I only have 10 actors. I remember I set this to 18 or something before. Idk why I changed it.

Comment (rkooo567) [reply python/ray/tests/test_multi_node_3.py:413]: I think 20 / 25 CPUs should be good enough to avoid flakiness? For more detailed analysis, I  think we can just do it inside a release test not here.

Comment (rkooo567) [reply python/ray/tests/test_multi_node_3.py:413]: I think 3 nodes (5CPUs each), 10 tasks,  2 min count is reasonably good? The only time it will fail is 5/5/0, and 5/4/1 cases. which is ((9)/10C3) == 10%? (lmk if the calculation is wrong lol)

Comment (wuisawesome) [reply python/ray/tests/test_multi_node_3.py:413]: Can we just use 2 nodes and a large number of cpus per node so it's much easier to reason about and less flakey?

If we do 2 nodes, 10 tasks and mincount=2, flakiness is ~0.6%.

Comment (rkooo567) [reply python/ray/tests/test_multi_node_3.py:413]: 2 nodes 10 tasks & mincount 2 will just always succeed right? because the cluster is fully utilized.

I think 10% is a good enough number because we retry 20 times.

**Review by wuisawesome — APPROVED** [review]

**Pull request merged**
